### ISSUE TITLE
Create Harvest DB and Table Structure

### ISSUE THREAD

**Issue by btylerburton** [GSA/data.gov#4612]
## User Story

In order to track the state of the new Harvesting 2.0 pipeline, datagovteam wants to create a Harvesting DB to record the status of harvest sources, harvest jobs, and harvest errors.

The Harvest DB will be defined in code using SQLAlchemy, so as to be reproducible through environments and deployments.

NOTE: As Google docs allows for more flexibility in table structure as modifications are ongoing, we will use the Harvesting Research Doc ([here](https://docs.google.com/document/d/1XzfTrPxu-asJ_55GoeZ2UOJsie9CuCegStS28BAL_40/edit#heading=h.b4usjhwwy0mv)) to finalize table structure before adding an accompanying ERD to `datagov-harvesting-logic` docs.

## Acceptance Criteria

[ACs should be clearly demoable/verifiable whenever possible. Try specifying them using [BDD](https://en.wikipedia.org/wiki/Behavior-driven_development#Behavioral_specifications).]

- [x] GIVEN the team has settled on the Table structure above
  AND it has been defined in an SQLAlchemy file
  THEN it should be able to be created programaticaly in a Docker container by running a DB initialization script.

- [x] GIVEN that the DB has been initialized
  WHEN tests are run locally in pytest
  THEN the results should be posted to the DB and be verifiable by running a query to generate the resulting report

## Background

[Any helpful contextual notes or links to artifacts/evidence, if needed]

## Security Considerations ([required](https://nvd.nist.gov/800-53/Rev4/control/CM-4))

[comment]: # "Our SSP says 'The Data.gov team ensures security implications are considered as part of the agile requirements refinement process by including a section in the issue template used as a basis for new work.' so please don't remove this section without care."
[Any security concerns that might be implicated in the change. "None" is OK, just be explicit here!]

- DB should be properly secured

## Sketch

- [x] Define Table Structure in SQLAlchemy
- [x] Initialize this structure into a local Docker PostGres container
- [x] Create an interface in datagov-harvesting-logic to interact with the DB
- [x] Call that interface as needed to report the results of harvester operations into the DB

**Comment by rshewitt** [comment]
[sqlachemy work](https://github.com/GSA/datagov-harvesting-logic/commit/9aab59f0c55fe7f0c93e2875b78c8bfe34696be7) in harvesting logic

**Comment by rshewitt** [comment]
[active branch](https://github.com/GSA/datagov-harvesting-logic/tree/define-sqlachemy)

**Comment by rshewitt** [comment]
[performance in sqlalchemy](https://docs.sqlalchemy.org/en/13/faq/performance.html)

**Comment by rshewitt** [comment]
[web request pattern in sqlalchemy](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#session-faq-whentocreate:~:text=the%20basic%20pattern%20is%20create%20a%20Session%20at%20the%20start%20of%20a%20web%20request%2C%20call%20the%20Session.commit()%20method%20at%20the%20end%20of%20web%20requests%20that%20do%20POST%2C%20PUT%2C%20or%20DELETE%2C%20and%20then%20close%20the%20session%20at%20the%20end%20of%20web%20request)

**Comment by rshewitt** [comment]
migration and db versioning links
- https://alembic.sqlalchemy.org/en/latest/
- https://docs.sqlalchemy.org/en/20/core/metadata.html#altering-database-objects-through-migrations

**Comment by Jin-Sun-tts** [comment]
Defined table structure in SQLAlchemy, create a docker container running Postgres, connected to create tables in DB:
![Screenshot 2024-02-13 at 12 55 56 PM](https://github.com/GSA/data.gov/assets/104456257/5f75283c-03af-4ac0-a387-6f23f6adf56c)

Inserted a sample record in the harvest_source table:
![Screenshot 2024-02-13 at 12 56 16 PM](https://github.com/GSA/data.gov/assets/104456257/ba72715b-eef2-47c0-9e6e-171207288187)

Created HarvestDBInterface to create/get/update db. Will add more detail functions in. 
also created a simple testing to insert records to the harvest_job table etc.

**Comment by robert-bryson** [comment]
DB created in dev space, see keys with `cf service-key harvesting-logic-db access-test`. Connecting to the service remotely needs to be done through an `ssh` tunnel: https://docs.cloudfoundry.org/devguide/deploy-apps/ssh-services.html#ssh-tunnel. I'm working on a proof of concept to connect to the internal from a github action.

**Comment by jbrown-xentity** [comment]
> DB created in dev space, see keys with `cf service-key harvesting-logic-db access-test`. Connecting to the service remotely needs to be done through an `ssh` tunnel: https://docs.cloudfoundry.org/devguide/deploy-apps/ssh-services.html#ssh-tunnel. I'm working on a proof of concept to connect to the internal from a github action.

Just FYI, we have ssh turned off on prod at the entire space level. So if we need to use ssh to connect, that's probably not a good long term approach... However, knowing that we cannot connect outside of cloud.gov to the RDS services is helpful to know!

**Comment by btylerburton** [comment]
I would think the long-term solution is a proxy app.

**Comment by btylerburton** [comment]
>Connecting to the service remotely needs to be done through an ssh tunnel

I tried to connect with SQLEctron yesterday and got basically the same connection error.

**Comment by Jin-Sun-tts** [comment]
Developed a Flask application and set up a PostgreSQL Docker container for local execution, as detailed in branch [create-harvest-db](https://github.com/GSA/datagov-harvesting-logic/tree/create-harvest-db). 

Additionally, interfaces were established to interact with the corresponding tables.

for examples, to insert test records, utilize the following endpoints:
Add a source:  /add_source
Add a job: /add_job?source_id=<id>
Add an error: /add_error?job_id=<id>

![Image](https://github.com/GSA/data.gov/assets/104456257/b8b29ae4-ec26-4015-9075-6f778ead84e7)

To-do:
Question: Regarding the harvest_source table, should organization_name and config.url be designated as the primary/unique key?
Add additional load test scripts 
Setup and conduct testing on cloud.gov, utilizing the database defined within that environment.

**Comment by rshewitt** [comment]
> Developed a Flask application and set up a PostgreSQL Docker container for local execution, as detailed in branch [create-harvest-db](https://github.com/GSA/datagov-harvesting-logic/tree/create-harvest-db).
> 
> Additionally, interfaces were established to interact with the corresponding tables.
> 
> for examples, to insert test records, utilize the following endpoints: Add a source: /add_source Add a job: /add_job?source_id= Add an error: /add_error?job_id=
> 
> ![Image](https://private-user-images.githubusercontent.com/104456257/305222294-b8b29ae4-ec26-4015-9075-6f778ead84e7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0NTMwNzYsIm5iZiI6MTcwODQ1Mjc3NiwicGF0aCI6Ii8xMDQ0NTYyNTcvMzA1MjIyMjk0LWI4YjI5YWU0LWVjMjYtNDAxNS05MDc1LTZmNzc4ZWFkODRlNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMFQxODEyNTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04YWViMzkwNzc5ZWJiYjkzYzI3ZjNkODJhZWIxMDIxODkxMjY1ZmUyOWI3ZTYxZGViM2QwMGJjZTFjNjUyNzZjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.PBmohYOVScLKR-ER7eiMCuJf0GwH49yCEy-krarSmy4)
> 
> To-do: Question: Regarding the harvest_source table, should organization_name and config.url be designated as the primary/unique key? Add additional load test scripts Setup and conduct testing on cloud.gov, utilizing the database defined within that environment.

@Jin-Sun-tts in response to your question, so far all the tables inherit an `id` field from `Base` ( [source](https://github.com/GSA/datagov-harvesting-logic/blob/22c380afa890378f6536daaeac38aee49776c15c/harvester/database/models.py#L6-L9 ) ) which is used as the primary key. are you suggesting creating a composite key using organization_name + config.url? I think that's more meaningful from a data perspective than using a `uuid`. I suppose it's a matter of whether we want to use a business key or a surrogate key.

**Comment by Jin-Sun-tts** [comment]
Pushed the Fask app to cloud.gov development  https://harvesting-logic.app.cloud.gov/

**Comment by btylerburton** [comment]
@Jin-Sun-tts chating with @krishnasanaka about the Flask form she's working on, and I believe there is work to be done to update the harvest_source table to the latest schema ([here](https://docs.google.com/document/d/1XzfTrPxu-asJ_55GoeZ2UOJsie9CuCegStS28BAL_40/edit)). 

```
Id (str) *primary
name (str)
notification_emails (array)
organization_name (str) (*self-reported)
frequency (str) (cron)
url (str)
schema_type (str)
source_type (str)
```

@GSA/data-gov-team can you take a look at the updates and add comments if you feel something was lost in translation? thanks.

https://docs.google.com/document/d/1XzfTrPxu-asJ_55GoeZ2UOJsie9CuCegStS28BAL_40/edit

**Comment by rshewitt** [comment]
> @Jin-Sun-tts chating with @krishnasanaka about the Flask form she's working on, and I believe there is work to be done to update the harvest_source table to the latest schema ([here](https://docs.google.com/document/d/1XzfTrPxu-asJ_55GoeZ2UOJsie9CuCegStS28BAL_40/edit)).
> 
> ```
> Id (str) *primary
> name (str)
> notification_emails (array)
> organization_name (str) (*self-reported)
> frequency (str) (cron)
> url (str)
> schema_type (str)
> source_type (str)
> ```
> 
> @GSA/data-gov-team can you take a look at the updates and add comments if you feel something was lost in translation? thanks.
> 
> https://docs.google.com/document/d/1XzfTrPxu-asJ_55GoeZ2UOJsie9CuCegStS28BAL_40/edit

we ended up exploding the `config` json field into separate fields so `url` and `schema_type` are their own thing now

**Comment by Jin-Sun-tts** [comment]
@btylerburton @rshewitt agree, that is also what @FuhuXia and I talked about the other day.  And we may have the `url` as unique key.

**Comment by Jin-Sun-tts** [comment]
Do we require the primary key `id` to be of type UUID, generated each time a record is added, OR retrieved/copied as a string from the CKAN database?

**Comment by btylerburton** [comment]
Removed AC/Sketch regarding load test, as that is dependent on #4617 and related to #4619. 

Copied old AC/Sketch below:

```
GIVEN that the above action has been successful locally
AND a database has been created on cloud.gov
THEN it should be possible to run a load test via Github Actions and generate a report of that test from the entries in the DB
```

`Invoke a load test and confirm that the expected results are written to the DB`

**Comment by rshewitt** [comment]
> Do we require the primary key `id` to be of type UUID, generated each time a record is added, OR retrieved/copied as a string from the CKAN database?

i figure `id` for harvest_error and harvest_job would be generated on insertion. using the ckan `id` in harvest_source allows us to relate tables by that field which could be valuable?

**Comment by btylerburton** [comment]
>  using the ckan id in harvest_source allows us to relate tables by that field which could be valuable?

Which `id` are we talking about? I believe CKAN would generate an identifier for a dataset, but would there be any concept of a harvest source in this new system? I don't think so.

**Comment by rshewitt** [comment]
@btylerburton I believe we're talking about the `id` field for harvest_source, harvest_error, and harvest_job ( [source](https://github.com/GSA/datagov-harvesting-logic/blob/bc14e288fddaba1bea6f9c489a8e43a6e11a9a5d/harvester/database/models.py#L6-L9) )

**Comment by btylerburton** [comment]
to add to harvest_source db table:
- harvest_source_id
- harvest_source_name

@krishnasanaka these are non-required text fields, copy/pasted over from CKAN

**Comment by Jin-Sun-tts** [comment]
I have drafted a pull request (PR) at https://github.com/GSA/datagov-harvesting-logic/pull/36 containing all the changes related to this issue. Need to transfer these changes to a separate repository, https://github.com/GSA/datagov-harvest-orchestrator, to facilitate further Flask UI modifications.

**Comment by Jin-Sun-tts** [comment]
@krishnasanaka Please reference the changes mentioned above and ensure that they are incorporated into the new repository `datagov-harvest-orchestrator`. Additionally, proceed with further modifications as required for ticket [#4619](https://github.com/GSA/data.gov/issues/4619)

**Comment by rshewitt** [comment]
we should think about things we want to seed into the database as part of the initialization

**Comment by Jin-Sun-tts** [comment]
@jbrown-xentity @rshewitt @btylerburton @FuhuXia  @robert-bryson 
Could you please share your opinion on using Flask-SQLAlchemy versus pure SQLAlchemy? 

Please let me know if it's necessary to create a new ticket for conducting research and experimentation with Flask-SQLAlchemy?

**Comment by jbrown-xentity** [comment]
> @jbrown-xentity @rshewitt @btylerburton @FuhuXia @robert-bryson Could you please share your opinion on using Flask-SQLAlchemy versus pure SQLAlchemy?
> 
> Please let me know if it's necessary to create a new ticket for conducting research and experimentation with Flask-SQLAlchemy?

I don't have any knowledge or opinion on the subject, discovery would be warranted unless someone else has experience...

**Comment by robert-bryson** [comment]
I don't have a strong opinion. I haven't used that specific project. Sometimes those intersectional type projects are very handy. Sometimes they are more hassle than they're worth. Sometimes they aren't maintained or the maintainer goes MIA without explanation.

**Comment by btylerburton** [comment]
This offers some good context. Overall, seems like global session management is the win, and locking us into that lib for better or worse is the price of that.

https://stackoverflow.com/questions/14343740/flask-sqlalchemy-or-sqlalchemy

**Comment by Jin-Sun-tts** [comment]
thank you everyone for the feedback. We will conduct further research and uncover additional insights within the Flask app ticket https://github.com/GSA/data.gov/issues/4619

Close this ticket as the table structure settled and defined in SQLAlchemy, and DB initialized and pytest tests run successfully, with results posted and verifiable via interfaces.
